File: functions/envioclick.js
const axios = require('axios');
const { findDaneCode, ORIGIN_CODE, ORIGIN_STATE } = require('./dane_codes');

const API_KEY = "6118c7fb-4b7d-4be9-92ce-503c55f40444";
const API_URL = "https://api.envioclickpro.com.co/api/v2";

const getHeaders = () => ({
    'Authorization': API_KEY, // Note: Docs say "key", check if "Bearer" needed. Docs example: "Authorization: 54ac..." (No Bearer)
    'Content-Type': 'application/json'
});

const ORIGIN_DATA = {
    "company": "Rostro Dorado Clinic",
    "firstName": "Rostro",
    "lastName": "Dorado",
    "email": "contacto@rostrodorado.com",
    "phone": "3000000000",
    "address": "Calle 12 #12-03 local 2",
    "suburb": "Centro",
    "crossStreet": "Calle 12",
    "reference": "Local 2",
    "daneCode": ORIGIN_CODE // Riohacha
};

/**
 * Quote Shipping Cost
 * Maps Envioclick response to our standard format.
 */
async function quoteShipping(data) {
    try {
        const { city, department, items, total } = data;
        console.log(`[Envioclick Quote] Request: ${city}, ${department}, Items: ${items?.length}, Total: ${total}`);

        const destination = findDaneCode(city, department);
        if (!destination) {
            console.error(`[Envioclick Quote] DANE Code NOT FOUND for: ${city}, ${department}`);
            return { success: false, error: `Ciudad no cubierta (${city}).` };
        }
        console.log(`[Envioclick Quote] DANE Match: ${destination.cityCode}`);

        // Calculate dimensional weight roughly if needed, or use provided dimensions
        // Envioclick requires packages array
        // FIX: Normalise weight (Grams -> Kg)
        const totalWeightRaw = items.reduce((acc, i) => acc + (i.weight || 500) * i.quantity, 0);
        const weightKg = totalWeightRaw > 10 ? totalWeightRaw / 1000 : totalWeightRaw;

        const packageData = {
            weight: Math.max(0.5, weightKg),
            height: 10, // Default or calculate max
            width: 10,
            length: 10
        };

        const payload = {
            "packages": [packageData],
            "description": "Productos de Belleza",
            "contentValue": total > 0 ? total : 20000,
            "origin": {
                "daneCode": ORIGIN_CODE,
                "address": ORIGIN_DATA.address
            },
            "destination": {
                "daneCode": destination.cityCode,
                "address": "Calle Principal" // Placeholder for quote
            }
        };

        const res = await axios.post(`${API_URL}/quotation`, payload, { headers: getHeaders() });
        // console.log("[Envioclick Quote] API Response Status:", res.status);

        if (res.data.status === 'OK' && res.data.data.rates) {
            console.log("RAW RATES EXTRACT:", JSON.stringify(res.data.data.rates.slice(0, 2), null, 2));
            // Map keys to match previous Envia format for Frontend compatibility
            const quotes = res.data.data.rates.map(rate => ({
                // Standard internal fields
                carrier: rate.carrier,
                service: rate.product,
                shippingCost: rate.flete,
                deliveryDay: rate.deliveryDays,

                // Envia-like structure for Frontend (CheckoutPage.tsx)
                deliveryCompany: {
                    companyName: rate.carrier,
                    deliveryEstimate: `${rate.deliveryDays} días hábiles`,
                    shippingCost: rate.flete,
                    service: rate.product
                },

                // Critical for Envioclick Generation
                idRate: rate.idRate,
                idProduct: rate.idProduct
            }));

            // Filter out non-viable options if needed
            return { success: true, quotes };
        } else {
            return { success: false, error: "No se encontraron cotizaciones." };
        }

    } catch (error) {
        console.error("[Envioclick Quote] API Error Payload:", JSON.stringify(error.response?.data || {}));
        console.error("Envioclick Quote Error Msg:", error.message);
        return { success: false, error: "Error al cotizar con Envioclick." };
    }
}

/**
 * Create Shipment
 * Uses idRate if fresh, or re-quotes if needed.
 */
async function createShipment(data) {
    try {
        const { customer, items, total, shippingOption } = data;

        const destination = findDaneCode(customer.city, customer.department);
        if (!destination) throw new Error("Invalid destination city.");

        // We assume shippingOption has idRate from the quote step.
        // But if it expired, we might fail. 
        // Strategy: Try with stored idRate. If fail, re-quote to get new idRate.

        let idRate = shippingOption?.idRate;

        // Construct Packages Object (Reusable)
        // Construct Packages Object (Reusable)
        // CRITICAL FIX: Convert weight from Grams to Kg.
        // If weight is > 10, assume it's Grams and divide. If < 10, assume it's already Kg.
        // This is a heuristic because DB has mixed units potentially.
        const totalWeightRaw = items.reduce((acc, i) => acc + (i.weight || 500) * i.quantity, 0);
        const weightKg = totalWeightRaw > 10 ? totalWeightRaw / 1000 : totalWeightRaw;

        const packageData = {
            weight: Math.max(0.5, weightKg), // Min 0.5 Kg
            height: 10,
            width: 10,
            length: 10
        };

        const buildShipmentPayload = (rateId) => ({
            "idRate": rateId,
            "myShipmentReference": `ORD-${Date.now()}`,
            "requestPickup": false,
            "insurance": true,
            "description": "Productos de Belleza",
            "contentValue": total > 0 ? total : 20000,
            "packages": [packageData],
            "origin": {
                ...ORIGIN_DATA,
                "firstName": "Rostro",
                "lastName": "Dorado",
            },
            "destination": {
                "company": "Particular", // Required min 2 chars, cannot be empty
                "firstName": customer.firstName,
                "lastName": customer.lastName,
                "email": customer.email,
                "phone": (customer.phone || '3000000000').replace(/\D/g, '').slice(-10) || '3000000000',
                "address": (customer.address || 'Calle Principal').slice(0, 40),
                "suburb": customer.neighborhood || "Centro", // Required field
                "crossStreet": "N/A",
                "reference": customer.notes || "Residencial",
                "daneCode": destination.cityCode
            }
        });

        // Try generation (Defaults to Sandbox for now, change URL for Prod)
        // CHECKOUT: Use `shipment_sandbox` for testing as per docs?
        // User wants to move to production eventually? 
        // Let's make it configurable. 
        // For now, let's assume PRODUCTION unless configured otherwise?
        // Actually, user gave up on Envia sandbox. 
        // Envioclick has specific sandbox endpoint: /shipment_sandbox

        const isSandbox = false; // PRODUCTION MODE
        const endpoint = isSandbox ? '/shipment_sandbox' : '/shipment';

        if (idRate) {
            try {
                const payload = buildShipmentPayload(idRate);
                console.log(`[CreateShipment] Attempting with ID: ${idRate}`);
                const res = await axios.post(`${API_URL}${endpoint}`, payload, { headers: getHeaders() });

                if (res.data.status === 'OK') {
                    return {
                        success: true,
                        trackingNumber: res.data.data.tracker,
                        labelUrl: res.data.data.url,
                        carrier: shippingOption?.carrier || "Envioclick"
                    };
                }
            } catch (e) {
                console.warn(`[CreateShipment] IdRate failed (${e.message}). Re-quoting...`);
            }
        }

        // Fallback: Re-Quote to get fresh idRate
        console.log("[CreateShipment] Re-quoting...");
        const quoteResult = await quoteShipping({
            city: customer.city,
            department: customer.department,
            items,
            total
        });

        if (quoteResult.success && quoteResult.quotes.length > 0) {
            const bestQuote = quoteResult.quotes.find(q => q.carrier === shippingOption?.carrier) || quoteResult.quotes[0];
            console.log(`[CreateShipment] New Quote ID: ${bestQuote.idRate} (${bestQuote.carrier})`);

            const payload = buildShipmentPayload(bestQuote.idRate);
            const res = await axios.post(`${API_URL}${endpoint}`, payload, { headers: getHeaders() });

            if (res.data.status === 'OK') {
                return {
                    success: true,
                    trackingNumber: res.data.data.tracker,
                    labelUrl: res.data.data.url,
                    carrier: bestQuote.carrier
                };
            }
        }

        throw new Error("Failed to generate shipment after retry.");

    } catch (error) {
        // ERROR HANDLING & PARSING
        const apiError = error.response?.data || {};
        let readableError = "Envioclick API Failed: Unknown error";

        // Try to extract specific message from Envioclick's nested structure
        // Structure: { status_messages: [ { error: [ "Unprocessed Entity", "No tiene suficiente crédito..." ] } ] }
        if (apiError.status_messages && Array.isArray(apiError.status_messages)) {
            const firstMsg = apiError.status_messages[0];
            if (firstMsg?.error && Array.isArray(firstMsg.error)) {
                // Join all error strings, e.g. "Unprocessed Entity. No tiene suficiente crédito..."
                readableError = firstMsg.error.join(' ');
            }
        } else if (apiError.message) {
            readableError = apiError.message;
        } else if (error.message) {
            readableError = error.message;
        }

        console.error("[CreateShipment] FINAL ERROR:", readableError);
        // Return success: false with the readable error so the Admin UI displays it
        return { success: false, error: readableError };
    }
}

/**
 * Track Shipment
 */
async function trackShipment(trackingCode) {
    try {
        const payload = { "trackingCode": trackingCode };
        // Post to /track
        const res = await axios.post(`${API_URL}/track`, payload, { headers: getHeaders() });

        if (res.data.status === 'OK' && res.data.data) {
            return {
                success: true,
                status: res.data.data.status,
                detail: res.data.data.statusDetail
            };
        }
        return { success: false, error: "Tracking not found" };
    } catch (error) {
        console.error("Envioclick Track Error:", error.response?.data || error.message);
        return { success: false, error: error.message };
    }
}

module.exports = { quoteShipping, createShipment, trackShipment };

File: backend/envioclick.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"regexp"
	"strings"
	"time"
)

const DefaultApiKey = "6118c7fb-4b7d-4be9-92ce-503c55f40444"
const ApiUrl = "https://api.envioclickpro.com.co/api/v2"

func getApiKey() string {
	if key := os.Getenv("ENVIOCLICK_API_KEY"); key != "" {
		return key
	}
	return DefaultApiKey
}

type PackageData struct {
	Weight float64 `json:"weight"`
	Height float64 `json:"height"`
	Width  float64 `json:"width"`
	Length float64 `json:"length"`
}

type OriginData struct {
	Company     string `json:"company"`
	FirstName   string `json:"firstName"`
	LastName    string `json:"lastName"`
	Email       string `json:"email"`
	Phone       string `json:"phone"`
	Address     string `json:"address"`
	Suburb      string `json:"suburb"`
	CrossStreet string `json:"crossStreet"`
	Reference   string `json:"reference"`
	DaneCode    string `json:"daneCode"`
}

var OriginInfo = OriginData{
	Company:     "Rostro Dorado Clinic",
	FirstName:   "Rostro",
	LastName:    "Dorado",
	Email:       "contacto@rostrodorado.com",
	Phone:       "3000000000",
	Address:     "Calle 12 #12-03 local 2",
	Suburb:      "Centro",
	CrossStreet: "Calle 12",
	Reference:   "Local 2",
	DaneCode:    OriginCode,
}

// Request models

type QuoteItem struct {
	Weight   float64 `json:"weight"`
	Quantity int     `json:"quantity"`
}

type QuoteRequest struct {
	City       string      `json:"city"`
	Department string      `json:"department"`
	Items      []QuoteItem `json:"items"`
	Total      float64     `json:"total"`
}

type ShipmentCustomer struct {
	FirstName      string `json:"firstName"`
	LastName       string `json:"lastName"`
	Email          string `json:"email"`
	Phone          string `json:"phone"`
	Address        string `json:"address"`
	City           string `json:"city"`
	Department     string `json:"department"`
	Neighborhood   string `json:"neighborhood"`
	Notes          string `json:"notes"`
	Identification string `json:"identification"`
}

type CreateShipmentRequest struct {
	Customer       ShipmentCustomer `json:"customer"`
	Items          []QuoteItem      `json:"items"`
	Total          float64          `json:"total"`
	ShippingOption struct {
		IdRate    int    `json:"idRate"`
		IdProduct int    `json:"idProduct"`
		Carrier   string `json:"carrier"`
		Service   string `json:"service"`
	} `json:"shippingOption"`
}

// Response models

type Rate struct {
	Carrier      string  `json:"carrier"`
	Product      string  `json:"product"`
	Flete        float64 `json:"flete"`
	DeliveryDays int     `json:"deliveryDays"`
	IdRate       int     `json:"idRate"`
	IdProduct    int     `json:"idProduct"`
}

type EnvioclickResponseData struct {
	Rates   []Rate `json:"rates"`
	Tracker string `json:"tracker"`
	URL     string `json:"url"`
	Status  string `json:"status"`
	Detail  string `json:"statusDetail"`
}

type EnvioclickResponse struct {
	Status         string                 `json:"status"`
	Data           EnvioclickResponseData `json:"data"`
	StatusMessages []struct {
		Error []string `json:"error"`
	} `json:"status_messages"`
	Message string `json:"message"`
}

type DeliveryCompanyInfo struct {
	CompanyName      string  `json:"companyName"`
	DeliveryEstimate string  `json:"deliveryEstimate"`
	ShippingCost     float64 `json:"shippingCost"`
	Service          string  `json:"service"`
}

type FrontendQuote struct {
	Carrier         string              `json:"carrier"`
	Service         string              `json:"service"`
	ShippingCost    float64             `json:"shippingCost"`
	DeliveryDay     int                 `json:"deliveryDay"`
	DeliveryCompany DeliveryCompanyInfo `json:"deliveryCompany"`
	IdRate          int                 `json:"idRate"`
	IdProduct       int                 `json:"idProduct"`
}

func doRequest(method, endpoint string, payload interface{}) ([]byte, error) {
	var bodyReader io.Reader
	if payload != nil {
		b, err := json.Marshal(payload)
		if err != nil {
			return nil, fmt.Errorf("marshal payload error: %w", err)
		}
		bodyReader = bytes.NewReader(b)
	}

	req, err := http.NewRequest(method, ApiUrl+endpoint, bodyReader)
	if err != nil {
		return nil, fmt.Errorf("create request error: %w", err)
	}

	req.Header.Set("Authorization", getApiKey())
	req.Header.Set("Content-Type", "application/json")

	client := &http.Client{Timeout: 15 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("http request error: %w", err)
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("read response body error: %w", err)
	}

	return respBody, nil
}

func QuoteShipping(reqData QuoteRequest, daneCodes []DaneCode) ([]FrontendQuote, error) {
	destination := FindDaneCode(reqData.City, reqData.Department, daneCodes)
	if destination == nil {
		return nil, fmt.Errorf("ciudad no cubierta (%s, %s)", reqData.City, reqData.Department)
	}

	// Calculate total weight in Kg
	var totalWeightRaw float64
	for _, item := range reqData.Items {
		w := item.Weight
		if w == 0 {
			w = 500 // default 500g
		}
		totalWeightRaw += w * float64(item.Quantity)
	}

	weightKg := totalWeightRaw
	if totalWeightRaw > 10 {
		weightKg = totalWeightRaw / 1000 // Convert grams to kg if logic suggests it was in grams
	}
	if weightKg < 0.5 {
		weightKg = 0.5 // Min 0.5 Kg
	}

	pkg := PackageData{
		Weight: weightKg,
		Height: 10,
		Width:  10,
		Length: 10,
	}

	val := reqData.Total
	if val <= 0 {
		val = 20000 // default content value
	}

	payload := map[string]interface{}{
		"packages":     []PackageData{pkg},
		"description":  "Productos de Belleza",
		"contentValue": val,
		"origin": map[string]string{
			"daneCode": OriginCode,
			"address":  OriginInfo.Address,
		},
		"destination": map[string]string{
			"daneCode": destination.CityCode,
			"address":  "Calle Principal",
		},
	}

	respBytes, err := doRequest("POST", "/quotation", payload)
	if err != nil {
		return nil, fmt.Errorf("api request failed: %w", err)
	}

	var evResp EnvioclickResponse
	if err := json.Unmarshal(respBytes, &evResp); err != nil {
		return nil, fmt.Errorf("parse response failed: %w", err)
	}

	if evResp.Status != "OK" || len(evResp.Data.Rates) == 0 {
		return nil, fmt.Errorf("no se encontraron cotizaciones")
	}

	var quotes []FrontendQuote
	for _, r := range evResp.Data.Rates {
		quotes = append(quotes, FrontendQuote{
			Carrier:      r.Carrier,
			Service:      r.Product,
			ShippingCost: r.Flete,
			DeliveryDay:  r.DeliveryDays,
			DeliveryCompany: DeliveryCompanyInfo{
				CompanyName:      r.Carrier,
				DeliveryEstimate: fmt.Sprintf("%d días hábiles", r.DeliveryDays),
				ShippingCost:     r.Flete,
				Service:          r.Product,
			},
			IdRate:    r.IdRate,
			IdProduct: r.IdProduct,
		})
	}

	return quotes, nil
}

type CreatedShipment struct {
	TrackingNumber string `json:"trackingNumber"`
	LabelUrl       string `json:"labelUrl"`
	Carrier        string `json:"carrier"`
}

func CreateShipment(reqData CreateShipmentRequest, daneCodes []DaneCode) (*CreatedShipment, error) {
	destination := FindDaneCode(reqData.Customer.City, reqData.Customer.Department, daneCodes)
	if destination == nil {
		return nil, fmt.Errorf("destino no válido (%s)", reqData.Customer.City)
	}

	var totalWeightRaw float64
	for _, item := range reqData.Items {
		w := item.Weight
		if w == 0 {
			w = 500
		}
		totalWeightRaw += w * float64(item.Quantity)
	}

	weightKg := totalWeightRaw
	if totalWeightRaw > 10 {
		weightKg = totalWeightRaw / 1000
	}
	if weightKg < 0.5 {
		weightKg = 0.5
	}

	pkg := PackageData{
		Weight: weightKg,
		Height: 10,
		Width:  10,
		Length: 10,
	}

	val := reqData.Total
	if val <= 0 {
		val = 20000
	}

	// Format phone: keep only last 10 digits
	reg := regexp.MustCompile(`\D`)
	phoneClean := reg.ReplaceAllString(reqData.Customer.Phone, "")
	if len(phoneClean) > 10 {
		phoneClean = phoneClean[len(phoneClean)-10:]
	}
	if phoneClean == "" {
		phoneClean = "3000000000"
	}

	addr := reqData.Customer.Address
	if len(addr) > 40 {
		addr = addr[:40]
	}

	suburb := reqData.Customer.Neighborhood
	if suburb == "" {
		suburb = "Centro"
	}

	ref := reqData.Customer.Notes
	if ref == "" {
		ref = "Residencial"
	}

	buildPayload := func(rateId int) map[string]interface{} {
		return map[string]interface{}{
			"idRate":               rateId,
			"myShipmentReference":  fmt.Sprintf("ORD-%d", time.Now().UnixNano()/1e6),
			"requestPickup":        false,
			"insurance":            true,
			"description":          "Productos de Belleza",
			"contentValue":         val,
			"packages":             []PackageData{pkg},
			"origin":               OriginInfo,
			"destination": map[string]interface{}{
				"company":     "Particular",
				"firstName":   reqData.Customer.FirstName,
				"lastName":    reqData.Customer.LastName,
				"email":       reqData.Customer.Email,
				"phone":       phoneClean,
				"address":     addr,
				"suburb":      suburb,
				"crossStreet": "N/A",
				"reference":   ref,
				"daneCode":    destination.CityCode,
			},
		}
	}

	endpoint := "/shipment" // PRODUCTION MODE
	idRate := reqData.ShippingOption.IdRate

	if idRate > 0 {
		payload := buildPayload(idRate)
		respBytes, err := doRequest("POST", endpoint, payload)
		if err == nil {
			var evResp EnvioclickResponse
			if json.Unmarshal(respBytes, &evResp) == nil && evResp.Status == "OK" {
				return &CreatedShipment{
					TrackingNumber: evResp.Data.Tracker,
					LabelUrl:       evResp.Data.URL,
					Carrier:        reqData.ShippingOption.Carrier,
				}, nil
			}
		}
	}

	// Fallback: Re-quote
	fmt.Println("[Envioclick Go] Re-quoting because idRate failed or was missing...")
	quoteReq := QuoteRequest{
		City:       reqData.Customer.City,
		Department: reqData.Customer.Department,
		Items:      reqData.Items,
		Total:      reqData.Total,
	}

	quotes, err := QuoteShipping(quoteReq, daneCodes)
	if err != nil || len(quotes) == 0 {
		return nil, fmt.Errorf("failed to re-quote: %v", err)
	}

	// Find best quote matching preferred carrier or pick the first
	bestQuote := quotes[0]
	for _, q := range quotes {
		if strings.EqualFold(q.Carrier, reqData.ShippingOption.Carrier) {
			bestQuote = q
			break
		}
	}

	payload := buildPayload(bestQuote.IdRate)
	respBytes, err := doRequest("POST", endpoint, payload)
	if err != nil {
		return nil, fmt.Errorf("re-shipment API request failed: %w", err)
	}

	var evResp EnvioclickResponse
	if err := json.Unmarshal(respBytes, &evResp); err != nil {
		return nil, fmt.Errorf("parse re-shipment response failed: %w", err)
	}

	if evResp.Status == "OK" {
		return &CreatedShipment{
			TrackingNumber: evResp.Data.Tracker,
			LabelUrl:       evResp.Data.URL,
			Carrier:        bestQuote.Carrier,
		}, nil
	}

	// Parse custom nested errors from Envioclick
	var errorMsg = "Envioclick API Failed: Unknown error"
	if len(evResp.StatusMessages) > 0 {
		if len(evResp.StatusMessages[0].Error) > 0 {
			errorMsg = strings.Join(evResp.StatusMessages[0].Error, " ")
		}
	} else if evResp.Message != "" {
		errorMsg = evResp.Message
	}

	return nil, fmt.Errorf(errorMsg)
}

type TrackResult struct {
	Status string `json:"status"`
	Detail string `json:"detail"`
}

func TrackShipment(trackingCode string) (*TrackResult, error) {
	payload := map[string]string{"trackingCode": trackingCode}
	respBytes, err := doRequest("POST", "/track", payload)
	if err != nil {
		return nil, fmt.Errorf("api request failed: %w", err)
	}

	var evResp EnvioclickResponse
	if err := json.Unmarshal(respBytes, &evResp); err != nil {
		return nil, fmt.Errorf("parse response failed: %w", err)
	}

	if evResp.Status == "OK" {
		return &TrackResult{
			Status: evResp.Data.Status,
			Detail: evResp.Data.Detail,
		}, nil
	}

	return nil, fmt.Errorf("tracking code not found")
}
